mlir/test/Dialect/Tosa/constant-cast-opt.mlir

```diff
@@ -0,0 +1,242 @@
+func.func @cast_fold_f32_to_i1_all_none_zero() -> tensor<3xi1> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}true{{.*}}tensor<3xi1>
+  // CHECK-NOT: tosa.cast
+  // CHECK: return [[RES]]
+  %0 = "tosa.const"() {value = dense<[12.0, 4.0, 5.0]> : tensor<3xf32>} : () -> tensor<3xf32>
+  %1 = "tosa.cast"(%0) : (tensor<3xf32>) -> tensor<3xi1>
+  return %1 : tensor<3xi1>
+}
+
+// CHECK-LABEL: @cast_fold_f32_to_i1
+func.func @cast_fold_f32_to_i1() -> tensor<3xi1> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}true, false, true{{.*}}tensor<3xi1>
+  // CHECK-NOT: tosa.cast
+  // CHECK: return [[RES]]
+  %0 = "tosa.const"() {value = dense<[12.0, 0.0, 5.0]> : tensor<3xf32>} : () -> tensor<3xf32>
+  %1 = "tosa.cast"(%0) : (tensor<3xf32>) -> tensor<3xi1>
+  return %1 : tensor<3xi1>
+}
+
+// CHECK-LABEL: @cast_fold_f32_to_i32
+func.func @cast_fold_f32_to_i32() -> tensor<3xi32> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}12, 4, 5{{.*}}tensor<3xi32>
+  // CHECK-NOT: tosa.cast
+  // CHECK: return [[RES]]
+  %0 = "tosa.const"() {value = dense<[12.0, 4.0, 5.0]> : tensor<3xf32>} : () -> tensor<3xf32>
+  %1 = "tosa.cast"(%0) : (tensor<3xf32>) -> tensor<3xi32>
+  return %1 : tensor<3xi32>
+}
+
+// CHECK-LABEL: @cast_fold_f32_to_i16
+func.func @cast_fold_f32_to_i16() -> tensor<5xi16> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}12, 0, 5, 32767, -32768{{.*}}tensor<5xi16>
+  // CHECK-NOT: tosa.cast
+  // CHECK: return [[RES]]
+  %0 = "tosa.const"() {value =
+                        dense<[12.0, 0.0, 5.0, 32770.11, -32770.11]> :
+                        tensor<5xf32>
+                      } : () -> tensor<5xf32>
+  %1 = "tosa.cast"(%0) : (tensor<5xf32>) -> tensor<5xi16>
+  return %1 : tensor<5xi16>
+}
+
+// CHECK-LABEL: @cast_fold_f16_to_i32
+func.func @cast_fold_f16_to_i32() -> tensor<6xi32> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}12, 14, 0, 5, 277, -278{{.*}}tensor<6xi32>
+  // CHECK-NOT: tosa.cast
+  // CHECK: return [[RES]]
+  %0 = "tosa.const"() {value =
+                        dense<[12.5, 13.5, 0.0, 5.0, 277.11, -277.71]> :
+                        tensor<6xf16>
+                      } : () -> tensor<6xf16>
+  %1 = "tosa.cast"(%0) : (tensor<6xf16>) -> tensor<6xi32>
+  return %1 : tensor<6xi32>
+}
+
+// CHECK-LABEL: @cast_fold_f32_to_i8
+func.func @cast_fold_f32_to_i8() -> tensor<5xi8> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}12, 0, 5, 127, -128{{.*}}tensor<5xi8>
+  // CHECK-NOT: tosa.cast
+  // CHECK: return [[RES]]
+  %0 = "tosa.const"() {value =
+                        dense<[12.0, 0.0, 5.0, 32770.11, -32770.11]> :
+                        tensor<5xf32>
+                      } : () -> tensor<5xf32>
+  %1 = "tosa.cast"(%0) : (tensor<5xf32>) -> tensor<5xi8>
+  return %1 : tensor<5xi8>
+}
+
+// CHECK-LABEL: @cast_fold_float_to_int_infinity_zero_nan
+func.func @cast_fold_float_to_int_infinity_zero_nan() -> tensor<5xi16> {
+  // CHECK: [[RES:]] ={{.*}}tosa.const{{.*}}32767, -32768, 0, 0{{.*}}tensor<5xi16>
```

Comment:
What is the result of 0x7FC00000?

Reply:
This is the special NaN to int case which is unspecified. I fold this to zero, however any value would be fine, therefore I did not check it. Should I put zero here?

Reply:
Comment added in 0222e16af672721482f05f11cdbdf9d5115e9639.